engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinObject.java

```diff
@@ -0,0 +1,93 @@
+package org.enso.interpreter.runtime.builtin;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.dsl.Bind;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import java.lang.ref.WeakReference;
+import org.enso.interpreter.node.expression.builtin.Builtin;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * Base class for every Enso builtin object. Not type. Note that base class for a builtin type is
+ * {@link Builtin}.
+ *
+ * <p>In other words, this class represents an object of builtin type in a similar way that {@link
+ * org.enso.interpreter.runtime.data.atom.Atom} represents an object of a non-builtin type.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+public abstract class BuiltinObject extends EnsoObject {
+
+  private final String builtinName;
+
+  /**
+   * A weak reference to the context in which this node was last executed.
+   *
+   * <p>Inspired by {@link
+   * org.enso.interpreter.node.expression.builtin.meta.EnsoProjectNode#previousCtxRef}
+   */
+  @CompilationFinal private WeakReference<EnsoContext> previousCtxRef = new WeakReference<>(null);
+
+  private Builtin cachedBuiltinType;
+
+  /**
+   * @param builtinName Simple name of the builtin that should be contained in {@link
+   *     org.enso.interpreter.runtime.builtin.Builtins#builtinsByName}.
+   */
+  protected BuiltinObject(String builtinName) {
+    this.builtinName = builtinName;
+  }
+
+  @ExportMessage
+  public final boolean hasType() {
+    return true;
+  }
+
+  @ExportMessage
+  public final Type getType(@Bind("$node") Node node) {
+    var ctx = EnsoContext.get(node);
```

Comment:
I don't like the `WeakReference` solution. It:
- would be good to find out who's creating the `BuiltinObject` instances and why they are shared between contexts?
- would be good to ensure each `EnsoContext` gets _own instances_ of `getBuiltins`
- if it turns out we have to have shared _builtin instances_, then change this to polymorphic `@Cached` specializing on `ctx`

Reply:
The latest revision on 7969907da39bcb886220b8a3c102f1943c08cd52 does not use `WeakReference` anymore.